File: app/history_worker.py
from __future__ import annotations

import json
import os
import time
from pathlib import Path

from .history import init_history, sync_print_history

APP_DIR = Path(os.getenv("APP_DATA", "/data"))
SETTINGS_FILE = APP_DIR / "settings.json"
DEFAULT_PRINTER_NAME = os.getenv("PRINTER_NAME", "Home_Epson_XP2200").strip() or "Home_Epson_XP2200"


def _positive_env_int(name: str, default: int) -> int:
    try:
        return max(1, int(os.getenv(name, str(default))))
    except (TypeError, ValueError):
        return default


POLL_INTERVAL_SECONDS = _positive_env_int("HISTORY_POLL_SECONDS", 5)
COMPLETED_POLL_SECONDS = max(POLL_INTERVAL_SECONDS, _positive_env_int("HISTORY_COMPLETED_POLL_SECONDS", 60))


def current_printer_name() -> str:
    try:
        data = json.loads(SETTINGS_FILE.read_text(encoding="utf-8"))
        value = str(data.get("printer_name", "")).strip()
        return value or DEFAULT_PRINTER_NAME
    except (OSError, json.JSONDecodeError):
        return DEFAULT_PRINTER_NAME


def main() -> None:
    init_history()
    print("[history] Persistent print history collector started.", flush=True)
    last_completed_poll = 0.0
    while True:
        try:
            now = time.monotonic()
            include_completed = now - last_completed_poll >= COMPLETED_POLL_SECONDS
            sync_print_history(current_printer_name(), include_completed=include_completed)
            if include_completed:
                last_completed_poll = now
        except Exception as exc:
            print(f"[history] Sync failed: {exc}", flush=True)
        time.sleep(POLL_INTERVAL_SECONDS)


if __name__ == "__main__":
    main()

File: src/history_worker.ts
import { join } from "node:path";
import { initHistory, syncPrintHistory } from "./history.ts";

const APP_DIR = process.env.APP_DATA || "/data";
const SETTINGS_FILE = join(APP_DIR, "settings.json");
const DEFAULT_PRINTER_NAME = (process.env.PRINTER_NAME || "Home_Epson_XP2200").trim() || "Home_Epson_XP2200";

function positiveEnvInt(name: string, def: number): number {
  const raw = process.env[name];
  if (raw === undefined) return def;
  const n = Number.parseInt(raw, 10);
  if (Number.isNaN(n)) return def;
  return Math.max(1, n);
}

export const POLL_INTERVAL_SECONDS = positiveEnvInt("HISTORY_POLL_SECONDS", 5);
export const COMPLETED_POLL_SECONDS = Math.max(POLL_INTERVAL_SECONDS, positiveEnvInt("HISTORY_COMPLETED_POLL_SECONDS", 60));

export async function currentPrinterName(): Promise<string> {
  try {
    const text = await Bun.file(SETTINGS_FILE).text();
    const data = JSON.parse(text);
    const value = String(data.printer_name ?? "").trim();
    return value || DEFAULT_PRINTER_NAME;
  } catch {
    return DEFAULT_PRINTER_NAME;
  }
}

export async function main(): Promise<void> {
  initHistory();
  console.log("[history] Persistent print history collector started.");
  let lastCompletedPoll = -COMPLETED_POLL_SECONDS;
  let running = false;
  while (true) {
    const tickStart = Date.now();
    try {
      if (!running) {
        running = true;
        try {
          const now = Date.now() / 1000;
          const includeCompleted = now - lastCompletedPoll >= COMPLETED_POLL_SECONDS;
          const name = await currentPrinterName();
          await syncPrintHistory(name, { includeCompleted });
          if (includeCompleted) lastCompletedPoll = now;
        } finally {
          running = false;
        }
      }
    } catch (exc: any) {
      running = false;
      console.log(`[history] Sync failed: ${exc?.message ?? String(exc)}`);
    }
    // account for sync duration so a slow CUPS fetch can't cause overlap/drift
    const elapsed = Date.now() - tickStart;
    await Bun.sleep(Math.max(1000, POLL_INTERVAL_SECONDS * 1000 - elapsed));
  }
}

if (import.meta.main) {
  main();
}
